### ISSUE TITLE
train file

### ISSUE THREAD

**Issue by Cai-RS** [antabangun/coex#16]
Thank you for your work! I'm a SLAM beginner and want to try using your work in a project, but I know very little about deep learning. Could you please provide the train file? I will be extremely grateful!

**Comment by antabangun** [comment]
Hi, thanks for your interest in our work!

The training scripts are in the file [stereo.py](https://github.com/antabangun/coex/blob/main/stereo.py). If you are interested in training the model, you can follow the README, especially the [dataset preparation](https://github.com/antabangun/coex#datasets). Then running `python stereo.py` should start the training, provided the environment has been configured correctly.

But if you only want to use the model in your SLAM system, I would suggest you use the provided model weights file and follow the codes in [demo.py](https://github.com/antabangun/coex/blob/main/demo.py) on how to use it.

**Comment by Cai-RS** [comment]
> Hi, thanks for your interest in our work!
> 
> The training scripts are in the file [stereo.py](https://github.com/antabangun/coex/blob/main/stereo.py). If you are interested in training the model, you can follow the README, especially the [dataset preparation](https://github.com/antabangun/coex#datasets). Then running `python stereo.py` should start the training, provided the environment has been configured correctly.
> 
> But if you only want to use the model in your SLAM system, I would suggest you use the provided model weights file and follow the codes in [demo.py](https://github.com/antabangun/coex/blob/main/demo.py) on how to use it.

Thanks for your reply and suggestions! According to my understanding, each dataset should be collected under different stereo camera configurations. Therefore, the network parameters obtained from different dataset reflect the different configurations of the stereo camera (i.e. camera extrinsic parameters). So if I want to test my SLAM system on other datasets such as EuRoC, Oxford, etc., do I have to retrain the network? Or does it just need fine-tuning?

**Comment by antabangun** [comment]
Ah, yes, you are correct about that. If the extrinsic configs are similar, the provided weights will perform better. Ideally, it would be best if you finetuned the network on the target dataset. But I think you could try using the provided weights first to avoid the pain of fine-tuning.

**Comment by Cai-RS** [comment]
> Ah, yes, you are correct about that. If the extrinsic configs are similar, the provided weights will perform better. Ideally, it would be best if you finetuned the network on the target dataset. But I think you could try using the provided weights first to avoid the pain of fine-tuning.

Okay, thanks for you advice. I would first try to test with the given parameters. In addition, did you use TensorRT for acceleration in the inference file? Is the 27ms/f shown in the demonstration under using of tensorRT?

**Comment by antabangun** [comment]
That reported time is without TensorRT. For TensorRT conversion you can refer to #8.

**Comment by Cai-RS** [comment]
> That reported time is without TensorRT. For TensorRT conversion you can refer to #8.

Thanks a lot for your patient reply! !
I have one last question, how does this network deal with the parts of the rectified image pair where the fields of view do not overlap? Or is it that the original image pairs are guaranteed to completely overlap? Because depth needs to be calculated later, it is necessary to know whether there are pixels that do not have valid disparity.

**Comment by antabangun** [comment]
Yes, assuming that we are using the left image as the reference image, some parts on the left side of the left image do not have the corresponding overlap on the right image. Most of the time, the trained network can do a decent job extrapolating predictions around these regions even without overlap. I could use them without any issues usually, especially if the target data is similar to the training data. But maybe not so much if the test data differs a lot. What I sometimes do in this situation is perform photometric consistency check (right shift followed by color and SSIM losses) around the regions that potentially do not have overlap, and discard disparity estimates with photometric loss above a certain threshold (see right most image in [this example](https://www.mdpi.com/sensors/sensors-21-07715/article_deploy/html/images/sensors-21-07715-g006-550.jpg)).

**Comment by Cai-RS** [comment]
Sincerely thank you for your guidance to me as a newbie! ! I'm sorry for replying to you so late, because I have been very busy recently, and I am also reading your paper on this work carefully. Now I know how to post-process the disparity map. But I have other questions. The paper mentions "We randomly crop images to size W = 576, H = 288 for training", but this is quite different from the resolution of the rectified images in the KITTI dataset (1242 & 375).  In addition to estimating depth, I also want to do object segmentation. So for the pre-processed stereo image I want to preserve as much content of the original image as possible. Using the cropping method will lose a lot of image area. Do you think it is feasible that I use resize (scaling + padding) during inference to get the specified input size (576*288) and then upsample and resize the disparity map to the resolution of the object segmentation network result?  Will it affect the depth estimation accuracy significantly? Or do you think I should change the structure of the network so that it can handle other resolution images (retraining should be troublesome?)?

Looking forward to hearing your suggestions.

**Comment by antabangun** [comment]
The random cropping to _(576*288)_ was just for data augmentation during training, which was commonly done for stereo model training to avoid overfitting and large training memory. You don't need to have the images to be that resolution during inference. 

The only thing the network expects is to have the input image resolutions be divisible by 32, and this is already handled in the `forward` function of `Stereo` class in [stereo.py](https://github.com/antabangun/coex/blob/main/stereo.py)
```
def forward(self, imgL, imgR=None, training=False):
        if training:
            disp_pred = self.stereo(imgL, imgR=imgR, training=training)
        else:
            h, w = imgL.shape[-2:]
            h_pad = 32-h % 32
            w_pad = 32-w % 32

            imgL = F.pad(imgL, (0, w_pad, 0, h_pad))
            if imgR is not None:
                imgR = F.pad(imgR, (0, w_pad, 0, h_pad))

            disp_pred = self.stereo(imgL, imR=imgR, training=training)[0][:, :h, :w]
            
        return disp_pred
```
So you can in fact input any image resolution to the model!

**Comment by Cai-RS** [comment]
After seeing your explanation, I found that there are networks such as FCN, which can allow the input image size to change... My knowledge of neural networks only stops at image classification tasks and fully connected layer!... You do help me a lot. Thanks sincerely again!

**Comment by Cai-RS** [comment]
Sorry to bother you again! When I was reading your preprocessing file, I noticed that you didn't seem to reproject the color image pairs of KITTI to ensure that the images came from cameras with coplanar projection planes. But according to my understanding of the rectified matrices in KITTI's calibration file, the four images of the rectified dataset in each timestamp are only rectified so that the camera projection planes are all parallel. Among them, only the projection plane of the two grayscale cameras (0 and 1) are coplanar, and the projection plane of the two color cameras are only parallel.

**Comment by antabangun** [comment]
Could you point me to which part of the preprocessing file you were referring to. 
I don't recall seeing anything about that on KITTI calibrations, so could you maybe also direct me to that?

**Comment by Cai-RS** [comment]
> Could you point me to which part of the preprocessing file you were referring to. I don't recall seeing anything about that on KITTI calibrations, so could you maybe also direct me to that?

 I think the rectifying process should be in all "***Loader.py" files in package "dataloaders/stereo/"，but I can't find any code about it in these files.  
To calculate the depth of a point based on disparity (the difference in x-coordinates of the projection of a certain 3D point on the left and right images), you must ensure that the image pair comes from cameras with their projection planes coplanar (so that there will be no disparity in the y direction) . However, in fact, KITTI's stereo dataset does not state that all rectified data represent that the projection planes of the four cameras are coplanar. What I did was extrapolate from the data form in the official calibration file.

**Comment by Cai-RS** [comment]
Since the real 3D coordinates of the optical centers of the four cameras are impossible to be coplanar (this can be seen from the calibration data of the raw dataset). Even for the rectified dataset, we can see from projection matrixes P0-P3 and rotation rectified matrix of camera0  R0_rect in the given calibration file (I randomly pick one frame): 
![1](https://github.com/antabangun/coex/assets/92929904/9d2378e8-b9d8-4f62-8f88-332f460299db)
and the formula for projecting the global 3D point to the rectified camera images:
![2](https://github.com/antabangun/coex/assets/92929904/4d3da8ef-150e-4d66-9c56-12da61563a4b)

Using the rectified camera 0 coordinate system as a reference, the optical centers of camera 0 and camera 1 only have displacements in the x direction, while the optical centers of the two color cameras (2 and 3) have displacements in three directions relative to the optical centers of camera 0 (can be seen from the last column of each P matrix). The displacement between 0 - 2 camera and displacement between  0 - 3 camera are not equal.
So I think we should perform another coplanar rectify on the rectified color image pair given by KITTI based on the P2 and P3 matrices in the calibration file. In each of your ***loader.py files, there is a kitti_calib function, which calculates the baseline length of the two color cameras, but does not correct the two cameras based on the optical center position. In fact, we know that the projection planes of the two cameras are parallel (no relative rotation), and the optical center displacement t and K matrix have been calculated in this function, then we only need to execute function like stereoRectify() in opencv and rectify can be completed.
![3](https://github.com/antabangun/coex/assets/92929904/3ba9baa4-ea69-4041-a570-880bc99cac7c)
However, it can be seen that the displacement values of the optical centers of camera 0 and cameras 2 and 3 are very small in the y and z directions, generally less than 5 mm. So maybe ignoring them won't make a difference?

**Comment by antabangun** [comment]
I see what you mean now, I believe you are correct. We didn't really pay attention to that when we wrote the codes because we didn't really use the calibration data for any part of the training. I don't know whether that 5mm can have a large impact on the reprojection or not. But feel free to submit a pull request for the fix

**Comment by Cai-RS** [comment]
> I see what you mean now, I believe you are correct. We didn't really pay attention to that when we wrote the codes because we didn't really use the calibration data for any part of the training. I don't know whether that 5mm can have a large impact on the reprojection or not. But feel free to submit a pull request for the fix

In fact, I don't know whether my inference is correct. But based on my knowledge of 3D vision and the specific form of calibration data, I can only draw this conclusion. However, I searched all day on the Internet and could not find any clear answer to this question. And in other stereo matching projects (such as the unofficial implementation of GOOGLE HITNET, the official implementation of PSMNET), I have not seen this problem dealt with. For example, the following code is from the unofficial HITNET implementation:
![4](https://github.com/antabangun/coex/assets/92929904/db888b63-e7b5-4080-b944-cab56dc62fed)
It directly defaults to the fact that the optical centers of the two cameras only have displacement in the x direction, so the calculated baseline length is rougher... In short, everyone seems to directly ignore the impact of this error. 
Anyway I don’t want to delve into this issue because I currently don’t have enough time and good equipment to retrain the network. I will use your trained network directly! Thanks a lot!
